### DISCUSSION TITLE
How does the parser distinguish between type aliases and primitive types?

### DISCUSSION THREAD

**Discussion by luan-xiaokun** [eclipse-langium/langium#904] [Q&A] [answered]
Hi guys, I've been playing with the Earley parser lately, and I found that the Earley parser (at least lark's earley implementation) treats type aliases (such as `typedef int as foo`) and primitive types (such as `int`, `bool`) as an ambiguity. So I start wondering how does langium (and many other LL/LALR parsers) distinguish between type alias and primitive types?

A snippet of the grammar:
```
ListType infers Type:
    (PrimitiveType | EnumType | StructType | AliasType | TupleType)
    ({infer ListType.base=current} '[' (capacity=Expression)? ']')*;
AliasType infers AliasType:
    alias=[NamedType:FeatureID];
FeatureID returns string:
    'value' | some_other_keywords | ID;
```
It makes sense to identify `int` as either a primitive type or an alias, as far as I can see.

My guess is: the keywords are collected by the parser generator and have a higher precedence than user-defined terminals.

Update: The question has been answered in #792 , keywords do take higher precedence.

**Comment by msujew** [comment, marked as answer]
_Just for other users who stumble upon this question_

As @luan-xiaokun correctly points out, in https://github.com/langium/langium/discussions/792 the precedence of keywords compared to terminals is explained. This allows correct lexing of keywords which would be otherwise identified as IDs or similar terminals. This behavior is implemented in the [`DefaultTokenBuilder`](https://github.com/langium/langium/blob/main/packages/langium/src/parser/token-builder.ts) class.

**Reply by luan-xiaokun** [reply]
Thank you for your reply, and sorry for not searching before asking.
